lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -239,14 +349,15 @@ def get_or_create_cached_value():
             except CacheKeyNotFoundError:
                 _LOGGER.debug("Cache miss: %s", func)
 
-                with cached_func.warning_call_stack.calling_cached_function(
-                    func
-                ), cached_func.message_call_stack.calling_cached_function():
-                    if cached_func.suppress_st_warning:
-                        with cached_func.warning_call_stack.suppress_cached_st_function_warning():
-                            return_value = func(*args, **kwargs)
-                    else:
-                        return_value = func(*args, **kwargs)
+                with cached_func.warning_call_stack.calling_cached_function(func):
+                    with cached_func.message_call_stack.calling_cached_function():
+                        with cached_func.message_call_stack.maybe_allow_widgets(
+                            cached_func.allow_widgets
+                        ):
+                            with cached_func.warning_call_stack.maybe_suppress_cached_st_function_warning(
+                                cached_func.suppress_st_warning
+                            ):
+                                return_value = func(*args, **kwargs)
```

Comment:
Nope, that's a python 3.8 (I think, might be 9) feature. That's how I originally wrote it, but it caused the 3.7 CI job to fail

Reply:
I think it may just be the parens that are python 3.9+? Earlier Pythons definitely support multiple context managers per `with` (we use the pattern in some of our tests, with some ugly line wrapping). (But it may be ugly enough to not be worthwhile!)

Reply:
Oh, right. I tried that before the parenthesized version and totally forgot about it. Something about python's parser requires the linebreaks to be in the middle of each expression, or else it won't parse, and the result is pretty bad.

```py
with cached_func.warning_call_stack.calling_cached_function(
func
), cached_func.message_call_stack.calling_cached_function(
), cached_func.message_call_stack.maybe_allow_widgets(
cached_func.allow_widgets
), cached_func.warning_call_stack.maybe_suppress_cached_st_function_warning(
cached_func.suppress_st_warning
):
    return_value = func(*args, **kwargs)
```
Something like this, though probably with slightly different indentation.